### PR TITLE
Wasp ai cli mage

### LINKED ISSUE #1575: Merge wasp-ai into main and improve Mage's CLI DX
- [x] Merge into `main`.
      What about the fact that `main` is going toward 0.12 version -> we should build in those change into Mage, right? Yeah, we should, we should update it.
- [x] Update Mage Wasp app to use newer version of Wasp?
- [x] Make CLI nice -> have it ask questions interactively, print stuff nicely, throw in some emojis ... . Remove debugging output.
- [x] Document CLI option better -> mention it in docs, maybe on the wasp landing page, certainly on the usemage.ai landing page, update FAQ there, add it to CLI usage docs, ... .
- [x] Document processes regarding deployment and hotfixes. If we want to make fixes to currently deployed version, we have to do it on `release`, don't we? What about `wasp-ai`?
- [x] Make it easy to provide an OpenAI key -> we can ask for it if you don't have it.
- [x] Possibly provide specifying any gpt model, not just the ones we have enumerated?

### PR DESCRIPTION
**Pull request by Martinsos**
This PR makes Wasp AI a proper, polished, well-documented feature of `wasp` CLI.

Resolves #1575 .

High-level list of changes:
1. If you choose `AI` in `wasp new`, you get asked a set of interactive questions about which GPT models and which temperature you want to use (besides app name and description).
2. `wasp new:ai` commands allows programmatically running Wasp AI. Its interface changed a bit, and it is now documented in the CLI usage output.
3. Output of Wasp AI from `wasp` CLI when writing to disk is now enriched with newlines, colors, and some extra information. In the meantime, output towards Mage hasn't changed (due to some cool logic we implemented in Haskell).
4. `wasp-ai/` dir got renamed to `mage/`.
5. Added README.md to `mage/`.
6. Updated all READMEs (/README.md, /waspc/README.md) with relevant info about mage/. Also updated docs.
7. Froze `wasp-ai` branch, from now on Mage will be deployed via `release` branch same as docs are deployed. Documented this in waspc/README.md.
8. Updated Mage to work with these newest changes.

Some screenshots of using Wasp AI via `wasp new`, interactively:

ai1
ai2
ai3

### CODE REVIEW

File: mage/README.md
Comment (Martinsos) [inline mage/README.md:44]:
```suggestion
A: When we say "Wasp AI" we refer to logic implemented in `wasp` CLI, while when we say "Mage" we refer to the Mage web app that really serves as a client for "Wasp AI" (calls it in the background). That said, we sometimes use these interchangeably.
```

File: waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates.hs
Comment (Martinsos) [inline waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates.hs:84]: @infomiho I made the change both here, and in Interactive, so that matches happen on prefix, not on the exact match, because I thought that will be more practical for users + it allowed to putting stuff like emojis at the end of the template title (one for AI).

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (Martinsos) [inline waspc/cli/src/Wasp/Cli/Interactive.hs:90]: I made it match by prefix, not just exact match @infomiho

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (Martinsos) [inline waspc/cli/src/Wasp/Cli/Interactive.hs:116]: I enriched the styling of options a bit, so now `(default)` tag is also yellow, and the index `[i]` of the default option is bold and underlined.

File: waspc/src/Wasp/AI/GenerateNewProject.hs
Comment (Martinsos) [inline waspc/src/Wasp/AI/GenerateNewProject.hs:48]: Here I added some extra info: I show GPT model and temeprature

File: waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs
Comment (Martinsos) [inline waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs:15]: This is our "Decorated String" type.

File: waspc/src/Wasp/AI/OpenAI/ChatGPT.hs
Comment (Martinsos) [inline waspc/src/Wasp/AI/OpenAI/ChatGPT.hs:56]: We had this printing all the time before. Now, I made it so that it prints only when DEBUG env vars is set.

File: waspc/src/Wasp/Util/Debug.hs
Comment (Martinsos) [inline waspc/src/Wasp/Util/Debug.hs:1]: OK I went a bit wild here, I added this "fake" pure function that checks if we are in debug mode.

I spent some time investigating other ways to accomplish running code conditionally depending on DEBUG env var, which included stuff like cabal flags and compiler pragmas, but all the solutions ended up boing overly complex, and this just seems in practice and at the moment to be the easiest. We can revisit it in the future but I would waste time trying to find something "cleaner". I did read online that doing unsafe lookup of env var like this shouldn't be able to cause any problems, so we should be ok.

File: waspc/src/Wasp/Util/Terminal.hs
Comment (Martinsos) [inline waspc/src/Wasp/Util/Terminal.hs:31]: Recent lesson about `fold`s educated me enough to fix this :D.

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (Martinsos) [inline waspc/cli/src/Wasp/Cli/Interactive.hs:54]: To learn about the motivation for this change, check comment https://github.com/wasp-lang/wasp/pull/1601/files#r1426920624 .

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:116]: Screenshot 2023-12-19 at 12 40 28

I'd remove the Bold and Underline for the index, it confused me for a second that I can move the options with my keyboard which I couldn't 😢  Until we have that, let's maybe not do anything different for the index. 

I like the yellow `(default)` 👍

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:90]: Aren't we afraid of users selecting the wrong thing sometimes e.g. 

- `What do you want to buy?`
- `car` vs. `carrot`

Type type in `car` but maybe get a `carrot`?

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (infomiho) [inline waspc/cli/src/Wasp/Cli/Interactive.hs:124]: I preferred the look of the description being aligned below the title 🤷 

Now it's a bit to the right 
Screenshot 2023-12-19 at 12 40 28

File: waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates.hs:84]: What if we have some sort of overlap with the names like `saas-minimal` and `saas-stripe`?

File: waspc/src/Wasp/Util/Terminal.hs
Comment (infomiho) [inline waspc/src/Wasp/Util/Terminal.hs:46]: This might not work in some terminal emulators 😢

File: waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates.hs
Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Command/CreateNewProject/StarterTemplates.hs:84]: Then it will take the first one, which I think is ok?

File: waspc/cli/src/Wasp/Cli/Command/CreateNewProject/AI.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Command/CreateNewProject/AI.hs:54]: Makes sense 👍 
`
One weird thing for me is the use of `askToChoose'` vs. `askToChoose` as the exposed function. Can we change that?

File: waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs
Comment (infomiho) [inline waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs:55]: `map` would make this more readable on first read 👍

File: waspc/src/Wasp/AI/CodeAgent.hs
Comment (infomiho) [reply waspc/src/Wasp/AI/CodeAgent.hs:38]: I don't understand the `logMsg` parameter, it has no constraints on it?

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:90]: Hm good point. They will get the first one that matches the prefix. But I haven't thought of this case, this is a problem! What should we do? Revert back? Or I can make it so it matches on exact match first, but if none, it matches by prefix?

Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:90]: Ok, I reverted it to as it was before.

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:116]: Good point, done!

File: waspc/src/Wasp/AI/CodeAgent.hs
Comment (Martinsos) [reply waspc/src/Wasp/AI/CodeAgent.hs:38]: So `CodeAgent` here is a standalone module, it could be in theory extracted in to a library -> it doesn't know anything about how we use it in Mage.

I could have made it directly use LogMsg data type which comes with it, but I don't really need to be that specific, instead I use anything that implements IsString -> that is enough for CogeAgent to know what to do with it. This in theory allows using CodeAgent with one type of log messages in one place, and with other kind of log message in some other use case.

Now, why there are no restrictions on the data type -> that is a bit non-instinctive in Haskell, but actually it is much better to put type constraints on functions that use the data type, not on data type itself. SO you will see I have `IsString` constraint for `logMsg` on some functions in this file. This has some benefits to it and is how it is done. If I am correct, putting constraints on the data type actually won't even have the wanted effect. We can discuss this more if you are interested into this topic.

File: waspc/src/Wasp/Util/Terminal.hs
Comment (Martinsos) [reply waspc/src/Wasp/Util/Terminal.hs:46]: I added a comment about that.

File: waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs
Comment (Martinsos) [reply waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs:55]: I find this more readable! Experience. But I suggest trying to get into it, I think once you get used to it, code gets easier to read. `<$>` is very common in Haskell, and you should get used to being able to read it quickly.

One thing I want to emphasize is that in Haskell, I most often don't bother understanding what every symbol does. I know compiler has my back and that stuff is "connected" correctly. What I care about is the general flow of data and what is happening.

So for example, when reading this code, I read it like this:
> ms is taken out of Concat, then toTermString is applied to it, and then concat is applied to that.

Notice that I don't actually care too much, at least in the first read, if it is `<$>` or `$`, I see it all as application. It doesn't really matter to me, I know that the thing on the right is now "term stringified" and that at the end all stuff is concatenated.

And when reading code like that, it is easier to read it when using `$` and `<$>` then when using `fmap` or `map` because they break my reading flow and force me to think about them.

Same is if I encounter an expression with a bunch of `>>=`, `<$>`, `$`, `>>`, `&`, ... .
Most of the time, I just need a general idea of what it does, and for that I don't need to understand exactly why each of those operators was used in that specific place. If I need to modify it, then I will dig a bit deeper, and again compiler will be helping me out.

File: waspc/cli/src/Wasp/Cli/Command/CreateNewProject/AI.hs
Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Command/CreateNewProject/AI.hs:54]: Makes sense, but both are valid choices to use -> one is for when dealing with Option, another is for more general case.

WHy not use `askToChoose` for all of them? Beacuse it would return `Option ...` and when you are using `Option`, you really want to get the value, not the whole option.

Since they are so similar, I thought `'` works fine ok, it is a specialized version of `askToChoose`.

**Comment by Martinsos** [comment]
@infomiho I responded to all comments and fixed stuff, pls give it another look!

File: waspc/cli/src/Wasp/Cli/Command/CreateNewProject/AI.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Command/CreateNewProject/AI.hs:54]: Okay, makes sense 👍

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:120]: Okay, it's obviously subjective, so I want insist on this 👍

**Comment by infomiho** [comment]
Looking good 👍

The indent change is a bit subjective, but I prefer the old style.

The `fmap` vs `map` is a bit subjective as well, so I won't press on that matter too hard.

File: waspc/cli/src/Wasp/Cli/Interactive.hs
Comment (infomiho) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:120]: I'd maybe look at other CLI tools how they indent their titles and descriptions. 

I found this in Astro's CLI 
Screenshot 2023-12-21 at 15 18 14

Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Interactive.hs:120]: Just a reminder, this is what it looks like with old indent:
image

This does look harder to read to me, because all text is so crammed up. In the example above from Astro, they have newlines, so that solves the issue, detaches title from the text. We don't have that here, nor would it be a good fit.

That said, since it is not obvious that change I made is better (both to you and @sodic ), then I will revert it to previous state.

Comment (Martinsos) [reply waspc/cli/src/Wasp/Cli/Interactive.hs]: Done!

**Comment by Martinsos** [comment]
> Looking good 👍
> 
> The indent change is a bit subjective, I prefer the old style 🤷
> 
> The `fmap` vs `map` is a bit subjective as well, so I won't press on that matter too hard.

Thanks! I reverted the indent change, and we agreed we will let people use what they want for map/fmap/<$>, so this should be it! If anything else let me know, otherwise I will merge once you approve.

**Review by infomiho — APPROVED** [review]
🚢  it!

File: waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs
Comment (infomiho) [reply waspc/src/Wasp/AI/GenerateNewProject/LogMsg.hs:55]: I still think that `map` is more easier to read for developers that have < 2 years of Haskell experience. But as we talked about it IRL, this seems a matter of personal preference and as I said already, I won't press this matter any more.

**Pull request merged**
